app/controllers/stats_controller.rb

```diff
@@ -34,31 +47,78 @@ def create_custom_report
     @report = Report.new(create_custom_params)
 
     if @report.valid?
-      @report.run(@report.period_start, @report.period_end)
+      if @report.xlsx?
+        report_data = @report.run(@report.period_start, @report.period_end)
+
+        axlsx = create_spreadsheet(report_data)
 
-      flash[:download] =  "Your custom report has been created. #{view_context.link_to 'Download', stats_download_custom_report_path(id: @report.id)}"
-      redirect_to stats_custom_path
+        send_data axlsx.to_stream.read,
+                  filename: @report.report_type.filename('xlsx'),
+                  disposition: :attachment,
+                  type: SPREADSHEET_CONTENT_TYPE
+      else
+        @report.run_and_update!(@report.period_start, @report.period_end)
+        flash[:download] =  "Your custom report has been created. #{view_context.link_to 'Download', stats_download_custom_report_path(id: @report.id)}"
+        redirect_to stats_custom_path
+      end
     else
       if create_custom_params[:correspondence_type].blank?
         @report.errors.add(:correspondence_type, :blank)
         @report.errors.delete(:report_type_id)
       end
-      @correspondence_types = CorrespondenceType.all
-      @custom_reports_foi = ReportType.custom.foi
-      @custom_reports_sar = ReportType.custom.sar
+      set_fields_for_custom_action
       render :custom
     end
   end
 
   def download_custom_report
     report= Report.find(params[:id])
-    filename = report.report_type.filename
+    filename = report.report_type.filename('csv')
 
     send_data report.report_data, {filename: filename, disposition: :attachment}
   end
 
   private
 
+  # The plan here is/was to colour the spreadsheet titles just like the existing reports from ITG
+  LIGHT_BLUE = 'c1d1f0'
+  LIGHT_GREY = 'd1d1e0'
+
+  BRIGHT_RED = 'FF0000'
+  BRIGHT_YELLOW = 'FFFF00'
+  BRIGHT_LIME_GREEN = '00FF00'
+
+  # mapping of RAG ratings to spreadsheet cell colours
+  RAG_RATING_COLOURS = { red: BRIGHT_RED, amber: BRIGHT_YELLOW, green: BRIGHT_LIME_GREEN }
+
+  # Assumes no report spans more than 26 columns
+  SPREADSHEET_COLUMN_NAMES = ('A'..'Z').to_a
+
```

Comment:
Is this the right place for these settings? Perhaps related to the comment above regarding `ExcelReport` class?

Reply:
Yes these might have been better in the view, but it's not easy to do given our current flow. It's more normal to know what your expected type is when you start the action - in 'custom report type' I don't think we know (unless maybe its always XLSX?) so the normal rails view pattern doesn't work